src/Hosting/TestHost/src/HttpContextBuilder.cs

```diff
@@ -59,15 +59,17 @@ internal void Configure(Action<HttpContext> configureContext)
         /// Start processing the request.
         /// </summary>
         /// <returns></returns>
-        internal Task<HttpContext> SendAsync(CancellationToken cancellationToken)
+        internal Task<HttpContext> SendAsync(CancellationToken cancellationToken, bool preserveExecutionContext)
         {
             var registration = cancellationToken.Register(AbortRequest);
 
-            _testContext = _application.CreateContext(_httpContext.Features);
-
-            // Async offload, don't let the test code block the caller.
-            _ = Task.Factory.StartNew(async () =>
+            // Everything inside this function happens in the SERVER's ExecutionContext (unless PreserveExecutionContext is true)
+            async Task RunRequestAsync()
             {
+                // This will configure IHttpContextAccessor so it needs to happen INSIDE this function,
+                // since we are now inside the Server's ExecutionContext. If it happens out
+                _testContext = _application.CreateContext(_httpContext.Features);
```

Comment:
We should capture and set the culture in here. This is the only common reliance on the execution context that I'm aware of.

Reply:
Hmm, for why?

Reply:
We use this in a bunch of places in tests where we're looking at something related to globlisation. That's the one dependency I think it's pretty reasonable to expect to work (setting culture).

Reply:
Why not just make the default True then so we don't need to special case anything?

Reply:
Yeah, it's a bit odd to special case one thing. The broken case (IHttpContextAccessor) is rare, but hard to troubleshoot. The culture case seems more common in our tests (not sure if any customers use this approach), but when it's not working I'd expect the cause to be more obvious since people are more familiar with how culture flows.

If culture is a thing you regularly want to test in an application why not make that a first class property of the TestServer? The server can set that in its own execution context.

Reply:
Yeah, I think that's an even better idea.

Reply:
I thought the culture setup in MVC sets it up on the middleware side, not the testserver side. https://github.com/aspnet/AspNetCore/blob/dcf49f257560130597b35f3b059b4ed34e8e3ead/src/Mvc/test/Mvc.FunctionalTests/Infrastructure/MvcTestFixture.cs#L21

https://github.com/aspnet/AspNetCore/blob/c565386a3ed135560bc2e9017aa54a950b4e35dd/src/Mvc/test/Mvc.FunctionalTests/Infrastructure/MvcWebApplicationBuilderExtensions.cs#L23

Reply:
Our tests don't seem to care what we do here, so I think this is more about existing **customer** scenarios.

Let's lay out the situation and come to a conclusion. This is my proposal based on this thread)

1. Default `PreserveExecutionContext` to `false` (I still think this is the default behavior users will want)
1. Add a `public CultureInfo RequestCuluture { get; set; }` property (feel free to bikeshed the name) to `TestServer`
1. If `RequestCulture != null`, set `CultureInfo.CurrentCulture` to that value on the "server side"
1. If `RequestCulture == null`, set `CultureInfo.CurrentCulture` on "the server" to whatever value it had on "the client" (this feels necessary to avoid breaking every loc-related test in user apps until they set `RequestCuluture`... or am I misunderstanding the use case?)

Reply:
Another question: If `PreserveExecutionContext == true` and `RequestCulture != null`, should we set `CultureInfo.CurrentCulture`? If we do, it will actually end up affecting the calling code, since we're in the same execution context.

Reply:
I wouldn't add the culture until people asked for it. Lets do nothing here until somebody asks. Today its already possible with middleware in the pipeline (or a startup filter).

Reply:
Ok with me. Our tests don't seem to be impacted at all by *not* flowing the culture. @rynowak ?

Reply:
I'm willing to just see what breaks then. If nothing broke when you tried this maybe we don't use it and we already do this in a more correct way.